packages/glimmer-reference/lib/iterable.ts

```diff
@@ -138,22 +143,27 @@ export class ReferenceIterator {
     this.artifacts = artifacts;
   }
 
-  next(): IterationItem<PathReference<Opaque>> {
+  getPosition(): UpdatableReference< number | FIXME<'user str to InternedString'>> {
+    return new UpdatableReference(this.iterator.getPosition());
+  }
+
+  next(): ListItem {
     let { artifacts } = this;
 
     let iterator = (this.iterator = this.iterator || artifacts.iterate());
 
     let item = iterator.next();
+    let position = iterator.getPosition();
 
     if (!item) return null;
 
-    return artifacts.append(item);
+    return artifacts.append(item, position);
```

Comment:
I think the reason why did this was that it didn't make sense to me that the position was part of the iterationItem interface since it was meta-item. Because of this I wound up passing everything across the stack. Given both your and Godfrey's comments, I'll make it internal to IterationItem so there's no more verbose passing around